File: datalib/src/main/java/nl/acidcats/tumblrlikes/data_impl/photodata/gateway/greendao/iterators/AbstractPhotoIterator.java
package nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.iterators;

import java.util.Iterator;

import nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.entities.PhotoEntity;
import nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.filters.FilterOption;

/**
 * Created on 17/10/2018.
 */
public abstract class AbstractPhotoIterator implements Iterator<PhotoEntity> {

    private int _size;
    private FilterOption _filterOption;

    public void setFilterOption(FilterOption filterOption) {
        _filterOption = filterOption;

        _size = filterOption.getCount();
    }

    @Override
    public boolean hasNext() {
        return true;
    }

    @Override
    public PhotoEntity next() {
        return _filterOption.getPhoto(getNextIndex());
    }

    protected abstract int getNextIndex();

    int getTotalCount() {
        return _size;
    }
}

File: datalib/src/main/java/nl/acidcats/tumblrlikes/data_impl/photodata/gateway/greendao/iterators/AbstractPhotoIterator.kt
package nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.iterators

import nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.entities.PhotoEntity
import nl.acidcats.tumblrlikes.data_impl.photodata.gateway.greendao.filters.FilterOption

/**
 * Created on 25/10/2018.
 */
abstract class AbstractPhotoIterator internal constructor(private val filterOption: FilterOption) : Iterator<PhotoEntity> {

    protected val totalCount by lazy { filterOption.count }

    protected abstract val nextIndex: Int

    override fun hasNext() = true

    override fun next(): PhotoEntity = filterOption.getPhoto(nextIndex)
}